WIP: Try to export subtree fields with attachments

renatgalimov opened the pull request:
Fixes #24 

Uses `org-export-as` with a region to see if that helps with preserving :DIR: attribute while exporting.

renatgalimov commented:
I will return to this PR next week.

renatgalimov commented on anki-editor.el:
```diff
@@ -793,13 +817,12 @@ Return a list of cons of (FIELD-NAME . FIELD-CONTENT)."
+                                      ;; e.g. by `org-map-entries' when
+                                      ;; scope is `tree'
+                                      (min (point-max) end)))
+                           (list 0 0))
```
```suggestion
                           "")
```

renatgalimov commented:
I got something useful here. Now I can export notes with attachments. @orgtre, any ideas about this PR?

renatgalimov commented:
This diff breaks `anki-editor--map-fields`, but I will deal with it later.

renatgalimov commented:
The whole idea worked pretty well.

But sometimes my code doesn't work properly in a big chunk of code called in `anki-editor--map-fields`[1]. It's a bit scary to try to fix it because the code is really big and complicated. Maybe it would be better to make it simpler first so it's easier to understand and find what's causing the problem.

 [1] https://github.com/orgtre/anki-editor/blob/ab7b33b48c78cb909391622eeffb893829b382d7/anki-editor.el#L891-L958

shunlog commented:
> But sometimes my code doesn't work properly in a big chunk of code called in `anki-editor--map-fields`

What exactly doesn't work?

renatgalimov commented:
@shunlog - I fixed the `anki-editor--map-fields`. 

Now, the note is exported not from a string but from a region, and the export is now aware about the org export context.

Also, I added a few unit tests to check how the export works. 

But I don't want to make more progress before @orgtre sees the diff.

slotThe reviewed (approved):
@renatgalimov is there anything you'd still like to do here? The code looks fine to me and the added tests are great!

slotThe commented:
N.b.: as someone who regularly pushes large subtrees of notes over to anki, I did notice a slowdown of at least one order of magnitude with this PR. I haven't profiled this yet, and I suspect it's fixable, but just something to keep in mind.

Since it fixes a bunch of outstanding issues (e.g., https://github.com/anki-editor/anki-editor/issues/38  and https://github.com/anki-editor/anki-editor/issues/39) the PR still seems worth merging, although one should definitely study the performance regression further

slotThe commented:
It seems like the reliance on `org-export-as` makes thing far from performant—plus, we spend over half of the time garbage collecting! Perhaps there was a reason the old code was manually assembling the strings for the fields.

@renatgalimov is it perhaps possible to keep `anki-editor--map-fields` as unchanged as possible, only fixing the problem in #24?

renatgalimov commented:
@slotThe, I can try to refactor it.

> @renatgalimov is there anything you'd still like to do here? The code looks fine to me and the added tests are great!

Thanks! I want to refactor the tests, as I don't like having a Python program to serve them. 

> It seems like the reliance on org-export-as makes thing far from performant—plus, we spend over half of the time garbage collecting! Perhaps there was a reason the old code was manually assembling the strings for the fields.

I can make it configurable, maintain old behavior by default, and enable the new one with a flag or setting. 

> @renatgalimov is it perhaps possible to keep anki-editor--map-fields as unchanged as possible

I will check it.
